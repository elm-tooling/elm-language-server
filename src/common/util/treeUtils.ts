import { Position } from "vscode-languageserver";
import { Node, Tree } from "web-tree-sitter";
import { ISourceFile } from "../../compiler/forest";
import { comparePosition, positionEquals } from "../positionUtil";
import { TRecord, Type } from "../../compiler/typeInference";
import { IProgram } from "../../compiler/program";
import {
  EFunctionCallExpr,
  mapSyntaxNodeToExpression,
} from "../../compiler/utils/expressionTree";
import { Range } from "vscode-languageserver-textdocument";
import { ISymbol } from "../../compiler/binder";

export type NodeType =
  | "Function"
  | "FunctionParameter"
  | "TypeAlias"
  | "Type"
  | "Operator"
  | "Module"
  | "CasePattern"
  | "AnonymousFunctionParameter"
  | "UnionConstructor"
  | "FieldType"
  | "TypeVariable"
  | "Port"
  | "Import";

const functionNameRegex = new RegExp("[a-zA-Z0-9_]+");

export class TreeUtils {
  public static getModuleNameNode(tree: Tree): Node | undefined {
    const moduleDeclaration: Node | undefined =
      this.findModuleDeclaration(tree);
    return moduleDeclaration?.childForFieldName("name") ?? undefined;
  }

  public static getModuleNameCommentNode(tree: Tree): Node | undefined {
    const moduleDeclaration: Node | undefined =
      this.findModuleDeclaration(tree);
    return moduleDeclaration?.nextNamedSibling?.type === "block_comment"
      ? moduleDeclaration.nextNamedSibling
      : undefined;
  }

  public static getModuleExposingListNodes(tree: Tree): Node[] {
    const moduleNode = TreeUtils.findModuleDeclaration(tree);

    if (moduleNode) {
      return [
        ...moduleNode.descendantsOfType(["exposed_value", "exposed_type"]),
      ];
    }

    return [];
  }

  public static findFirstNamedChildOfType(
    type: string,
    node: Node,
  ): Node | undefined {
    return node.children.find((child) => child.type === type);
  }

  public static findAllNamedChildrenOfType(
    type: string | string[],
    node: Node,
  ): Node[] | undefined {
    const result = Array.isArray(type)
      ? node.children.filter((child) => type.includes(child.type))
      : node.children.filter((child) => child.type === type);

    return result.length === 0 ? undefined : result;
  }

  public static findExposedFunctionNode(
    node: Node,
    functionName: string,
  ): Node | undefined {
    if (node) {
      const exposingList = this.findFirstNamedChildOfType(
        "exposing_list",
        node,
      );
      if (exposingList) {
        const doubleDot = this.findFirstNamedChildOfType(
          "double_dot",
          exposingList,
        );
        if (doubleDot) {
          return undefined;
        }
      }
      const descendants = TreeUtils.descendantsOfType(node, "exposed_value");
      return descendants.find((desc) => desc.text === functionName);
    }
  }

  public static isExposedFunctionOrPort(
    tree: Tree,
    functionName: string,
  ): boolean {
    const module = this.findModuleDeclaration(tree);
    if (module) {
      const exposingList = this.findFirstNamedChildOfType(
        "exposing_list",
        module,
      );
      if (exposingList) {
        const doubleDot = exposingList.childForFieldName("doubleDot");
        if (doubleDot) {
          return true;
        }
      }
      const descendants = TreeUtils.descendantsOfType(module, "exposed_value");
      return descendants.some((desc) => desc.text === functionName);
    }
    return false;
  }

  public static findExposedTypeOrTypeAliasNode(
    node: Node,
    typeName: string,
  ): Node | undefined {
    if (node) {
      const exposingList = this.findFirstNamedChildOfType(
        "exposing_list",
        node,
      );
      if (exposingList) {
        const doubleDot = this.findFirstNamedChildOfType(
          "double_dot",
          exposingList,
        );
        if (doubleDot) {
          return undefined;
        }
      }
      const descendants = TreeUtils.descendantsOfType(node, "exposed_type");
      const match = descendants.find(
        (desc) =>
          desc.text === typeName || `${desc.text}(`.startsWith(typeName),
      );
      if (match && match.firstNamedChild) {
        return match.firstNamedChild;
      }
    }
    return undefined;
  }

  public static isExposedTypeOrTypeAlias(
    tree: Tree,
    typeName: string,
  ): boolean {
    const module = this.findModuleDeclaration(tree);
    if (module) {
      const exposingList = this.findFirstNamedChildOfType(
        "exposing_list",
        module,
      );
      if (exposingList) {
        const doubleDot = this.findFirstNamedChildOfType(
          "double_dot",
          exposingList,
        );
        if (doubleDot) {
          return true;
        }
      }
      const typeNameDoubleDot = `${typeName}(..)`;
      const descendants = TreeUtils.descendantsOfType(module, "exposed_type");
      return descendants.some(
        (desc) => desc.text === typeName || desc.text === typeNameDoubleDot,
      );
    }
    return false;
  }

  public static findUnionConstructorCalls(
    tree: Tree,
    unionConstructorName: string,
    moduleNamePrefix?: string,
  ): Node[] | undefined {
    const upperCaseQid = TreeUtils.descendantsOfType(
      tree.rootNode,
      "upper_case_qid",
    );
    if (upperCaseQid.length > 0) {
      const result = upperCaseQid.filter(
        (a) =>
          (a.text === unionConstructorName ||
            (moduleNamePrefix &&
              a.text === `${moduleNamePrefix}.${unionConstructorName}`)) &&
          a.parent &&
          a.parent.type !== "type_ref" &&
          a.parent.type !== "import_clause",
      );
      return result.length === 0 ? undefined : result;
    }
  }

  public static findOperator(
    sourceFile: ISourceFile,
    operatorName: string,
  ): Node | undefined {
    const rootSymbols = sourceFile.symbolLinks?.get(sourceFile.tree.rootNode);

    const operatorNode = rootSymbols?.get(operatorName)?.node;

    if (operatorNode) {
      const functionReference = TreeUtils.findFirstNamedChildOfType(
        "value_expr",
        operatorNode,
      );
      if (functionReference) {
        return rootSymbols?.get(
          functionReference.text,
          (s) => s.node.type !== "infix_declaration",
        )?.node;
      }
    }
  }

  public static findTypeDeclaration(
    tree: Tree,
    typeName: string,
  ): Node | undefined {
    const types = this.findAllTypeDeclarations(tree);
    if (types) {
      return types.find(
        (a) =>
          a.children.length > 1 &&
          a.children[1].type === "upper_case_identifier" &&
          a.children[1].text === typeName,
      );
    }
  }

  public static findModuleDeclaration(tree: Tree): Node | undefined {
    return tree.rootNode.childForFieldName("moduleDeclaration") ?? undefined;
  }

  public static findAllTopLevelFunctionDeclarations(
    tree: Tree,
  ): Node[] | undefined {
    const result = tree.rootNode.children.filter(
      (a) => a.type === "value_declaration",
    );
    return result.length === 0 ? undefined : result;
  }

  public static getFunctionNameNodeFromDefinition(
    node: Node,
  ): Node | undefined {
    if (node.type === "lower_case_identifier") {
      return node;
    }
    const declaration =
      node.type == "function_declaration_left"
        ? node
        : node.childForFieldName("functionDeclarationLeft");
    if (declaration && declaration.firstNamedChild) {
      return declaration.firstNamedChild;
    }
  }

  public static getTypeOrTypeAliasOrPortNameNodeFromDefinition(
    node: Node,
  ): Node | undefined {
    return node.childForFieldName("name") ?? undefined;
  }

  public static isTypeUsage(upperCaseQid: Node): boolean {
    return (
      !!TreeUtils.findParentOfType("type_ref", upperCaseQid) ||
      upperCaseQid.parent?.type === "exposed_type"
    );
  }

  public static isConstructorUsage(upperCaseQid: Node): boolean {
    return upperCaseQid.parent?.type === "value_expr";
  }

  public static findTypeOrTypeAliasCalls(
    tree: Tree,
    typeOrTypeAliasName: string,
    nodeType: NodeType,
  ): Node[] {
    const upperCaseQids = TreeUtils.descendantsOfType(
      tree.rootNode,
      "upper_case_qid",
    );

    const supportsTypeUsage = nodeType === "Type" || nodeType === "TypeAlias";
    const supportsConstructorUsage = nodeType === "TypeAlias";

    return upperCaseQids.filter((a) => {
      return (
        a.text === typeOrTypeAliasName &&
        ((supportsTypeUsage && TreeUtils.isTypeUsage(a)) ||
          (supportsConstructorUsage && TreeUtils.isConstructorUsage(a)))
      );
    });
  }

  public static findAllTypeDeclarations(tree: Tree): Node[] | undefined {
    return this.findAllNamedChildrenOfType("type_declaration", tree.rootNode);
  }

  /**
   * @deprecated Should not be used due to performance. Use bindings instead
   */
  public static findImportClauseByName(
    tree: Tree,
    moduleName: string,
  ): Node | undefined {
    const allImports = this.findAllImportClauseNodes(tree);
    if (allImports) {
      return allImports.find(
        (a) =>
          a.children.length > 1 &&
          a.children[1].type === "upper_case_qid" &&
          a.children[1].text === moduleName,
      );
    }
  }

  public static getTypeOrTypeAliasOfFunctionParameter(
    node: Node | undefined,
  ): Node | undefined {
    if (
      node &&
      node.parent &&
      node.parent.parent &&
      node.parent.parent.parent &&
      node.parent.parent.parent.previousNamedSibling &&
      node.parent.parent.parent.previousNamedSibling.type ===
        "type_annotation" &&
      node.parent.parent.parent.previousNamedSibling.lastNamedChild
    ) {
      const functionParameterNodes = TreeUtils.findAllNamedChildrenOfType(
        ["pattern", "lower_pattern", "record_pattern"],
        node.parent.parent,
      );
      if (functionParameterNodes) {
        const matchIndex = functionParameterNodes.findIndex(
          (a) => a.text === node.text,
        );

        const typeAnnotationNodes = TreeUtils.findAllNamedChildrenOfType(
          ["type_ref", "type_expression"],
          node.parent.parent.parent.previousNamedSibling.lastNamedChild,
        );
        if (typeAnnotationNodes) {
          return typeAnnotationNodes[matchIndex];
        }
      }
    }
  }

  public static getReturnTypeOrTypeAliasOfFunctionDefinition(
    node: Node | undefined,
  ): Node | undefined {
    if (node && node.previousNamedSibling?.type === "type_annotation") {
      const typeAnnotationNodes = TreeUtils.descendantsOfType(
        node.previousNamedSibling,
        "type_ref",
      );
      if (typeAnnotationNodes) {
        const type = typeAnnotationNodes[typeAnnotationNodes.length - 1];
        return type.firstNamedChild?.firstNamedChild ?? type;
      }
    }
  }

  public static getRecordTypeOfFunctionRecordParameter(
    node: Node | undefined,
    program: IProgram,
  ): TRecord | undefined {
    const checker = program.getTypeChecker();
    if (
      node?.parent?.type === "function_call_expr" &&
      node.parent.firstNamedChild
    ) {
      const functionCallExpr = mapSyntaxNodeToExpression(
        node.parent,
      ) as EFunctionCallExpr;

      const parameterIndex =
        functionCallExpr.namedChildren.map((c) => c.text).indexOf(node.text) -
        1;

      const foundType = checker.findType(functionCallExpr.target);

      if (foundType.nodeType === "Function") {
        const paramType = foundType.params[parameterIndex];

        if (paramType.nodeType === "Record") {
          return paramType;
        }
      }
    }
  }

  public static getTypeAliasOfRecordField(
    node: Node | undefined,
    sourceFile: ISourceFile,
    program: IProgram,
  ): Node | undefined {
    const fieldName = node?.parent?.firstNamedChild?.text;

    let recordType = TreeUtils.getTypeAliasOfRecord(node, sourceFile, program);

    while (!recordType && node?.parent?.parent) {
      node = node.parent.parent;
      recordType = TreeUtils.getTypeAliasOfRecordField(
        node,
        sourceFile,
        program,
      );
    }

    const recordTypeSourceFile = recordType
      ? program.getSourceFile(recordType.tree.uri)
      : undefined;

    if (recordType && recordTypeSourceFile) {
      const fieldTypes = TreeUtils.descendantsOfType(recordType, "field_type");
      const fieldNode = fieldTypes.find((a) => {
        return (
          TreeUtils.findFirstNamedChildOfType("lower_case_identifier", a)
            ?.text === fieldName
        );
      });

      if (fieldNode) {
        const typeExpression = TreeUtils.findFirstNamedChildOfType(
          "type_expression",
          fieldNode,
        );

        if (typeExpression) {
          const typeNode = TreeUtils.descendantsOfType(
            typeExpression,
            "upper_case_identifier",
          );

          if (typeNode.length > 0) {
            const typeAliasNode = program
              .getTypeChecker()
              .findDefinition(typeNode[0], recordTypeSourceFile).symbol;

            if (typeAliasNode) {
              return typeAliasNode.node;
            }
          }
        }
      }
    }
  }

  public static getTypeAliasOfRecord(
    node: Node | undefined,
    sourceFile: ISourceFile,
    program: IProgram,
  ): Node | undefined {
    if (node?.parent?.parent) {
      let type: Node | undefined | null =
        TreeUtils.findFirstNamedChildOfType(
          "record_base_identifier",
          node.parent.parent,
        ) ??
        TreeUtils.findFirstNamedChildOfType(
          "record_base_identifier",
          node.parent,
        );

      // Handle records of function returns
      if (!type && node.parent.parent.parent) {
        type =
          TreeUtils.getReturnTypeOrTypeAliasOfFunctionDefinition(
            node.parent.parent.parent,
          )?.parent ?? undefined;
      }

      if (!type) {
        type = node;
      }

      if (type) {
        const definitionNode = program
          .getTypeChecker()
          .findDefinition(
            type.firstNamedChild ? type.firstNamedChild : type,
            sourceFile,
          ).symbol;

        if (definitionNode) {
          const definitionTree = program.getSourceFile(
            definitionNode.node.tree.uri,
          );

          let aliasNode;
          if (
            definitionNode.type === "FunctionParameter" &&
            definitionNode.node.firstNamedChild
          ) {
            aliasNode = TreeUtils.getTypeOrTypeAliasOfFunctionParameter(
              definitionNode.node.firstNamedChild,
            );
          } else if (definitionNode.type === "Function") {
            aliasNode = TreeUtils.getReturnTypeOrTypeAliasOfFunctionDefinition(
              definitionNode.node,
            );
          } else if (definitionNode.type === "FieldType") {
            aliasNode = TreeUtils.findFirstNamedChildOfType(
              "type_expression",
              definitionNode.node,
            );
          } else if (definitionNode.type === "TypeAlias") {
            return definitionNode.node;
          }

          if (aliasNode && definitionTree) {
            const childNode = TreeUtils.descendantsOfType(
              aliasNode,
              "upper_case_identifier",
            );

            if (childNode.length > 0) {
              const typeNode = program
                .getTypeChecker()
                .findDefinition(childNode[0], definitionTree).symbol;

              if (typeNode) {
                return typeNode.node;
              }
            }
          }
        }
      }
    }
  }

  public static getAllFieldsFromTypeAlias(
    node: Node | undefined,
  ): { field: string; type: string }[] | undefined {
    const result: { field: string; type: string }[] = [];
    if (node) {
      const fieldTypes = TreeUtils.descendantsOfType(node, "field_type");
      if (fieldTypes.length > 0) {
        fieldTypes.forEach((a) => {
          const fieldName = TreeUtils.findFirstNamedChildOfType(
            "lower_case_identifier",
            a,
          );
          const typeExpression = TreeUtils.findFirstNamedChildOfType(
            "type_expression",
            a,
          );
          if (fieldName && typeExpression) {
            result.push({ field: fieldName.text, type: typeExpression.text });
          }
        });
      }
    }
    return result.length === 0 ? undefined : result;
  }

  public static descendantsOfType(node: Node, type: string): Node[] {
    return node.descendantsOfType(type);
  }

  public static getNamedDescendantForPosition(
    node: Node,
    position: Position,
  ): Node {
    const previousCharColumn =
      position.character === 0 ? 0 : position.character - 1;
    const charBeforeCursor = node.text
      .split("\n")
      [position.line].substring(previousCharColumn, position.character);

    if (!functionNameRegex.test(charBeforeCursor)) {
      return node.namedDescendantForPosition({
        column: position.character,
        row: position.line,
      });
    } else {
      return node.namedDescendantForPosition(
        {
          column: previousCharColumn,
          row: position.line,
        },
        {
          column: position.character,
          row: position.line,
        },
      );
    }
  }

  public static getDescendantForPosition(node: Node, position: Position): Node {
    const previousCharColumn =
      position.character === 0 ? 0 : position.character - 1;
    const charBeforeCursor = node.text
      .split("\n")
      [position.line].substring(previousCharColumn, position.character);

    if (!functionNameRegex.test(charBeforeCursor)) {
      return node.descendantForPosition({
        column: position.character,
        row: position.line,
      });
    } else {
      return node.descendantForPosition(
        {
          column: previousCharColumn,
          row: position.line,
        },
        {
          column: position.character,
          row: position.line,
        },
      );
    }
  }

  public static getNamedDescendantForRange(
    sourceFile: ISourceFile,
    range: Range,
  ): Node {
    if (positionEquals(range.start, range.end)) {
      return this.getNamedDescendantForPosition(sourceFile.tree.rootNode, {
        character: range.start.character,
        line: range.start.line,
      });
    } else {
      return sourceFile.tree.rootNode.namedDescendantForPosition(
        {
          column: range.start.character,
          row: range.start.line,
        },
        {
          column: range.end.character,
          row: range.end.line,
        },
      );
    }
  }

  public static getDescendantForRange(
    sourceFile: ISourceFile,
    range: Range,
  ): Node {
    if (positionEquals(range.start, range.end)) {
      return this.getDescendantForPosition(sourceFile.tree.rootNode, {
        character: range.start.character,
        line: range.start.line,
      });
    } else {
      return sourceFile.tree.rootNode.descendantForPosition(
        {
          column: range.start.character,
          row: range.start.line,
        },
        {
          column: range.end.character,
          row: range.end.line,
        },
      );
    }
  }

  public static findPreviousNode(
    node: Node,
    position: Position,
  ): Node | undefined {
    function nodeHasTokens(n: Node): boolean {
      return n.endIndex - n.startIndex !== 0;
    }

    function findRightmostChildWithTokens(
      childrenList: Node[],
      startIndex: number,
    ): Node | undefined {
      for (let i = startIndex - 1; i >= 0; i--) {
        if (nodeHasTokens(childrenList[i])) {
          return childrenList[i];
        }
      }
    }

    function findRightmostNode(n: Node): Node | undefined {
      if (n.children.length === 0) {
        return n;
      }

      const candidate = findRightmostChildWithTokens(
        n.children,
        n.children.length,
      );

      if (candidate) {
        return findRightmostNode(candidate);
      }
    }

    const children = node.children;

    if (children.length === 0) {
      return node;
    }

    for (let i = 0; i < children.length; i++) {
      const child = children[i];
      if (comparePosition(position, child.endPosition) < 0) {
        const lookInPreviousChild =
          comparePosition(position, child.startPosition) <= 0 ||
          !nodeHasTokens(child);

        if (lookInPreviousChild) {
          const candidate = findRightmostChildWithTokens(children, i);
          if (candidate) {
            return findRightmostNode(candidate);
          }
        } else {
          return this.findPreviousNode(child, position);
        }
      }
    }

    const candidate = findRightmostChildWithTokens(children, children.length);
    if (candidate) {
      return findRightmostNode(candidate);
    }
  }

  public static getNamedDescendantForLineBeforePosition(
    node: Node,
    position: Position,
  ): Node {
    const previousLine = position.line === 0 ? 0 : position.line - 1;

    return node.namedDescendantForPosition({
      column: 0,
      row: previousLine,
    });
  }

  public static getNamedDescendantForLineAfterPosition(
    node: Node,
    position: Position,
  ): Node {
    const followingLine = position.line + 1;

    return node.namedDescendantForPosition({
      column: 0,
      row: followingLine,
    });
  }

  public static findParentOfType(
    typeToLookFor: string,
    node: Node,
    topLevel = false,
  ): Node | undefined {
    if (
      node.type === typeToLookFor &&
      (!topLevel || node.parent?.type === "file")
    ) {
      return node;
    }
    if (node.parent) {
      return this.findParentOfType(typeToLookFor, node.parent, topLevel);
    }
  }

  public static getLastImportNode(tree: Tree): Node | undefined {
    const allImportNodes = this.findAllImportClauseNodes(tree);
    if (allImportNodes?.length) {
      return allImportNodes[allImportNodes.length - 1];
    }
  }

  public static isReferenceFullyQualified(node: Node): boolean {
    return (
      node.previousNamedSibling?.type === "dot" &&
      node.previousNamedSibling?.previousNamedSibling?.type ===
        "upper_case_identifier"
    );
  }

  public static getTypeAnnotation(valueDeclaration?: Node): Node | undefined {
    if (valueDeclaration?.type !== "value_declaration") {
      return;
    }

    let candidate = valueDeclaration.previousNamedSibling;

    // Skip comments
    while (
      candidate?.type === "line_comment" ||
      candidate?.type === "comment_block"
    ) {
      candidate = candidate.previousNamedSibling;
    }

    if (candidate?.type === "type_annotation") {
      return candidate;
    }
  }

  public static getValueDeclaration(typeAnnotation?: Node): Node | undefined {
    if (typeAnnotation?.type !== "type_annotation") {
      return;
    }

    let candidate = typeAnnotation.nextNamedSibling;

    // Skip comments
    while (
      candidate?.type === "line_comment" ||
      candidate?.type === "comment_block"
    ) {
      candidate = candidate.nextNamedSibling;
    }

    if (candidate?.type === "value_declaration") {
      return candidate;
    }
  }

  /**
   * This gets a list of all ancestors of a type
   * in order from the closest declaration up to the top level declaration
   */
  public static getAllAncestorsOfType(type: string, node: Node): Node[] {
    const declarations = [];

    while (node.type !== "file") {
      if (node.type === type) {
        declarations.push(node);
      }

      if (node.parent) {
        node = node.parent;
      } else {
        break;
      }
    }

    return declarations;
  }

  /**
   * @deprecated Should not be used due to performance. Use bindings instead
   */
  public static findAllImportClauseNodes(tree: Tree): Node[] | undefined {
    const result = tree.rootNode.children.filter(
      (a) => a.type === "import_clause",
    );

    return result.length === 0 ? undefined : result;
  }

  public static isIdentifier(node: Node): boolean {
    return (
      node.type === "lower_case_identifier" ||
      node.type === "upper_case_identifier"
    );
  }

  public static isImport(node: Node): boolean {
    return (
      node.parent?.firstNamedChild?.type === "import" ||
      node.parent?.parent?.firstNamedChild?.type === "import"
    );
  }

  public static nextNode(node: Node): Node | undefined {
    // Move up until we have a sibling
    while (!node.nextNamedSibling && node.parent) {
      node = node.parent;
    }

    if (node.nextNamedSibling) {
      node = node.nextNamedSibling;

      // Move down the leftmost subtree
      while (node.firstNamedChild) {
        node = node.firstNamedChild;
      }

      return node;
    }
  }

  public static findFieldReference(
    type: Type,
    fieldName: string,
  ): ISymbol | undefined {
    if (type.nodeType === "Record") {
      const fieldRefs = type.fieldReferences.get(fieldName);

      if (fieldRefs.length > 0) {
        return { name: fieldName, node: fieldRefs[0], type: "FieldType" };
      }
    }
  }

  public static findImportAliasOfModule(
    moduleName: string,
    tree: Tree,
  ): string | undefined {
    const importClause = TreeUtils.findImportClauseByName(tree, moduleName);

    const asClause = importClause?.childForFieldName("asClause");

    if (asClause) {
      return asClause.childForFieldName("name")?.text;
    } else {
      return importClause?.childForFieldName("moduleName")?.text;
    }
  }
}
