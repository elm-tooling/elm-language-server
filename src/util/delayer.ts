/**
 * Delayer taken from VS Code typescript-language-features: https://github.com/microsoft/vscode/blob/a36c68b9ec3d6a0aca9799d7a10be741a6658a51/extensions/typescript-language-features/src/utils/async.ts#L10
 */

export interface ITask<T> {
  (): T;
}

export class Delayer<T> {
  public defaultDelay: number;
  private timeout: NodeJS.Timeout | null; // Timer
  private completionPromise: Promise<T | null> | null;
  private onSuccess: ((value: T | PromiseLike<T> | undefined) => void) | null;
  private task: ITask<T> | null;

  constructor(defaultDelay: number) {
    this.defaultDelay = defaultDelay;
    this.timeout = null;
    this.completionPromise = null;
    this.onSuccess = null;
    this.task = null;
  }

  public trigger(
    task: ITask<T>,
    delay: number = this.defaultDelay,
  ): Promise<T | null> {
    this.task = task;
    if (delay >= 0) {
      this.cancelTimeout();
    }

    if (!this.completionPromise) {
      this.completionPromise = new Promise<T | undefined>((resolve) => {
        this.onSuccess = resolve;
      }).then(() => {
        this.completionPromise = null;
        this.onSuccess = null;
        const result = this.task && this.task();
        this.task = null;
        return result;
      });
    }

    if (delay >= 0 || this.timeout === null) {
      this.timeout = setTimeout(
        () => {
          this.timeout = null;
          if (this.onSuccess) {
            this.onSuccess(undefined);
          }
        },
        delay >= 0 ? delay : this.defaultDelay,
      );
    }

    return this.completionPromise;
  }

  private cancelTimeout(): void {
    if (this.timeout !== null) {
      clearTimeout(this.timeout);
      this.timeout = null;
    }
  }
}
