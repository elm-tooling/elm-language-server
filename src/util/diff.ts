import diff from "fast-diff";
import { Connection, Range, TextEdit } from "vscode-languageserver";
import { URI } from "vscode-uri";
import { execCmdSync } from "../compiler/utils/elmUtils";
// Given two strings (`before`, `after`), return a list of all substrings
// that appear in `after` but not in `before`, and the positions of each
// of the substrings within `after`.
function getTextRangeChanges(
  before: string,
  after: string,
): Promise<TextEdit[]> {
  const newRanges: TextEdit[] = [];
  let lineNumber = 0;
  let column = 0;

  const parts = diff(before, after);

  // Loop over every part, keeping track of:
  // 1. The current line no. and column in the `after` string
  // 2. Character ranges for all "added" parts in the `after` string
  parts.forEach((part) => {
    const startLineNumber = lineNumber;
    const startColumn = column;
    if (part[0] === 0 || part[0] === -1) {
      // Split the part into lines. Loop through these lines to find
      // the line no. and column at the end of this part.
      const lines = part[1].split("\n").map((a) => a.length);
      lines.forEach((lineLength, lineIndex) => {
        // The first `line` is actually just a continuation of the last line
        if (lineIndex === 0) {
          column += lineLength;
          // All other lines come after a line break.
        } else if (lineIndex > 0) {
          lineNumber += 1;
          column = lineLength;
        }
      });
    }

    if (part[0] === 1) {
      newRanges.push({
        newText: part[1],
        range: Range.create(
          startLineNumber,
          startColumn,
          startLineNumber,
          startColumn,
        ),
      });
    } else if (part[0] === -1) {
      newRanges.push({
        newText: "",
        range: Range.create(startLineNumber, startColumn, lineNumber, column),
      });
    }
  });
  return Promise.resolve(newRanges);
}

export function formatText(
  elmWorkspaceRootPath: URI,
  elmFormatPath: string,
  text: string,
  connection: Connection,
): Promise<TextEdit[]> {
  const options = {
    cmdArguments: ["--stdin", "--elm-version", "0.19", "--yes"],
    notFoundText: "Install elm-format via 'npm install -g elm-format'.",
  };

  const format = execCmdSync(
    elmFormatPath,
    "elm-format",
    options,
    elmWorkspaceRootPath.fsPath,
    connection,
    text,
  );
  return getTextRangeChanges(text, format.stdout);
}
