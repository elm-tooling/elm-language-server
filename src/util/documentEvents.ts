import { EventEmitter } from "events";
import {
  DidChangeTextDocumentParams,
  DidCloseTextDocumentParams,
  DidOpenTextDocumentParams,
  DidSaveTextDocumentParams,
  Connection,
} from "vscode-languageserver";
import { injectable, container } from "tsyringe";

type DidChangeCallback = (params: DidChangeTextDocumentParams) => void;
type DidCloseCallback = (params: DidCloseTextDocumentParams) => void;
type DidOpenCallback = (params: DidOpenTextDocumentParams) => void;
type DidSaveCallback = (params: DidSaveTextDocumentParams) => void;

export interface IDocumentEvents {
  on(event: "change", listener: DidChangeCallback): this;
  on(event: "close", listener: DidCloseCallback): this;
  on(event: "open", listener: DidOpenCallback): this;
  on(event: "save", listener: DidSaveCallback): this;
}

@injectable()
export class DocumentEvents extends EventEmitter implements IDocumentEvents {
  constructor() {
    const connection = container.resolve<Connection>("Connection");
    super();

    connection.onDidChangeTextDocument((e) => this.emit("change", e));
    connection.onDidCloseTextDocument((e) => this.emit("close", e));
    connection.onDidOpenTextDocument((e) => this.emit("open", e));
    connection.onDidSaveTextDocument((e) => this.emit("save", e));
  }
}
