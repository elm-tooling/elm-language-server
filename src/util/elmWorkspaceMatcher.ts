import { container } from "tsyringe";
import { CancellationToken } from "vscode-languageserver";
import { URI } from "vscode-uri";
import { IProgram } from "../compiler/program";
import { ISourceFile } from "../compiler/forest";
import { NoWorkspaceContainsError } from "./noWorkspaceContainsError";

export interface IParams {
  program: IProgram;
  sourceFile: ISourceFile;
}

/**
 * Identifies the relevant ElmWorkspace for a given ParamType, either directly
 * (getElmWorkspaceFor) or when an event handler receives a ParamType
 * (handle) it returns a params object with a combined type ParamType and IParams
 * which has the program and sourceFile.
 */
export class ElmWorkspaceMatcher<ParamType> {
  private elmWorkspaces: IProgram[];

  constructor(private getUriFor: (param: ParamType) => URI) {
    this.elmWorkspaces = container.resolve("ElmWorkspaces");
  }

  public handle<ResultType>(
    handler: (
      param: ParamType & IParams,
      token?: CancellationToken,
    ) => ResultType,
  ): (param: ParamType, token?: CancellationToken) => ResultType {
    return (param: ParamType, token?: CancellationToken): ResultType => {
      const program = this.getProgramFor(param);
      return handler(
        {
          ...param,
          program,
          sourceFile: this.getSourceFileFor(param, program),
        },
        token,
      );
    };
  }

  public handleResolve<ResultType>(
    handler: (
      param: ParamType,
      program: IProgram,
      sourceFile: ISourceFile,
      token?: CancellationToken,
    ) => ResultType,
  ): (param: ParamType, token?: CancellationToken) => ResultType {
    return (param: ParamType, token?: CancellationToken): ResultType => {
      const program = this.getProgramFor(param);
      return handler(
        param,
        program,
        this.getSourceFileFor(param, program),
        token,
      );
    };
  }

  public getProgramFor(param: ParamType): IProgram {
    const uri = this.getUriFor(param);
    const program =
      // first look for a program where the file has been parsed to a tree
      this.elmWorkspaces.find((ws) => ws.hasDocument(uri)) ||
      // fallback: find a program where the file is in the source-directories
      this.elmWorkspaces.find((ws) => ws.isInSourceDirectory(uri.fsPath));

    if (!program) {
      throw new NoWorkspaceContainsError(this.getUriFor(param));
    }

    return program;
  }

  public getSourceFileFor(param: ParamType, program: IProgram): ISourceFile {
    const uri = this.getUriFor(param).toString();

    return program.getForest().getByUri(uri)!;
  }
}
