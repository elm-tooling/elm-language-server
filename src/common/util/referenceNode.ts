import { Node } from "web-tree-sitter";
import { NodeType } from "./treeUtils";

export interface IReferenceNode {
  node: Node;
  nodeType: NodeType;
  uri: string;
}
