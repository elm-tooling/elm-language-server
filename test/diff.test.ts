import "reflect-metadata";
import { container } from "tsyringe";
import { URI } from "vscode-uri";
import { formatText } from "../src/util/diff";
import { Connection } from "vscode-languageserver";

container.register("Connection", {
  useValue: {
    console: {
      info: (a: string): void => {
        // console.log(a);
      },
      warn: (a: string): void => {
        // console.log(a);
      },
      error: (a: string): void => {
        // console.log(a);
      },
    },
    window: {
      showErrorMessage: (a: string): void => {
        console.log(a);
      },
    },
  },
});
describe("test formatting", () => {
  const connection = container.resolve<Connection>("Connection");
  const pathUri = URI.file(__dirname);

  test("normal format gives correct result", async () => {
    const result = await formatText(
      pathUri,
      "elm-format",
      `
module Main

import Browser
import Html exposing (Html, button, div, text)
import Html.Events exposing (onClick)


type alias Model =
    { count : Int }


initialModel : Model
initialModel =
    { count = 0 }


type Msg
    = Increment
    | Decrement


update : Msg -> Model -> Model
update msg model =
    case msg of
        Increment ->
            { model | count = model.count + 1 }

        Decrement ->
            { model | count = model.count - 1 }


view : Model -> Html Msg
view model =
    div []
        [ button [ onClick Increment ] [ text "+1" ]
        , div [] [ text <| String.fromInt model.count ]
        , button [ onClick Decrement ] [ text "-1" ]
        ]


main : Program () Model Msg
main =
    Browser.sandbox
        { init = initialModel
        , view = view
        , update = update
        }
      `,
      connection,
    );

    expect(result).toMatchSnapshot();
  });
});
