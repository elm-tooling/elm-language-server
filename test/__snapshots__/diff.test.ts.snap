// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test formatting normal format gives correct result 1`] = `
Array [
  Object {
    "newText": "",
    "range": Object {
      "end": Object {
        "character": 0,
        "line": 1,
      },
      "start": Object {
        "character": 0,
        "line": 0,
      },
    },
  },
  Object {
    "newText": " exposing (Model, Msg(..), initialModel, main, update, view)",
    "range": Object {
      "end": Object {
        "character": 11,
        "line": 1,
      },
      "start": Object {
        "character": 11,
        "line": 1,
      },
    },
  },
  Object {
    "newText": "",
    "range": Object {
      "end": Object {
        "character": 6,
        "line": 48,
      },
      "start": Object {
        "character": 0,
        "line": 48,
      },
    },
  },
]
`;
